backend/app/notify/telegram.go

```diff
@@ -159,44 +187,129 @@ func escapeTelegramText(text string) string {
+func (t *Telegram) CheckToken(token, user string) (telegram, site string, err error) {
+	t.requests.RLock()
+	authRequest, ok := t.requests.data[token]
+	t.requests.RUnlock()
+
+	if !ok {
+		return "", "", errors.New("request is not found")
+	}
+
+	if time.Now().After(authRequest.expires) {
+		t.requests.Lock()
+		delete(t.requests.data, token)
+		t.requests.Unlock()
+		return "", "", errors.New("request expired")
 	}
-	select {
-	case <-ctx.Done():
-		return errors.Errorf("sending message to %q aborted due to canceled context", req.User)
-	default:
+
+	if !authRequest.confirmed {
+		return "", "", errors.New("request is not verified yet")
 	}
 
-	log.Printf("[DEBUG] send verification via %s, user %s", t, req.User)
-	msg, err := t.buildVerificationMessage(req.User, req.Token, req.SiteID)
-	if err != nil {
-		return err
+	if authRequest.user != user {
+		return "", "", errors.New("user does not match original requester")
 	}
 
-	return t.sendMessage(ctx, msg, req.Telegram)
+	// Delete request
+	t.requests.Lock()
+	delete(t.requests.data, token)
+	t.requests.Unlock()
+
+	return authRequest.telegramID, authRequest.site, nil
 }
 
-// buildVerificationMessage generates verification telegram message based on given input
-func (t *Telegram) buildVerificationMessage(user, token, site string) ([]byte, error) {
-	result := fmt.Sprintf("Confirmation for <i>%s</i> on site %s\n"+
-		"Please copy and paste this text into “token” field on comments page to confirm subscription:\n\n\n"+
-		"<pre>%s</pre>",
-		escapeTelegramText(user), escapeTelegramText(site), escapeTelegramText(token))
-	body := telegramMsg{Text: result, ParseMode: "HTML"}
-	b, err := json.Marshal(body)
-	if err != nil {
-		return nil, err
+// Run starts processing login requests sent in Telegram, required for user notifications to work
+// Blocks caller
+func (t *Telegram) Run(ctx context.Context) {
+	t.run = true
+	processUpdatedTicker := time.NewTicker(t.apiPollInterval)
+	cleanupTicker := time.NewTicker(t.expiredCleanupInterval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			processUpdatedTicker.Stop()
+			cleanupTicker.Stop()
+			return
+		case <-processUpdatedTicker.C:
+			updates, err := t.getUpdates(ctx)
+			if err != nil {
+				log.Printf("[WARN] Error while getting telegram updates: %v", err)
+				continue
+			}
+			t.processUpdates(ctx, updates)
+		case <-cleanupTicker.C:
+			now := time.Now()
+			t.requests.Lock()
+			for key, req := range t.requests.data {
+				if now.After(req.expires) {
+					delete(t.requests.data, key)
+				}
+			}
+			t.requests.Unlock()
+		}
 	}
-	return b, nil
 }
 
-// GetBotUsername returns bot username
-func (t *Telegram) GetBotUsername() string {
-	return t.username
+// ProcessUpdate is alternative to Run, it processes provided plain text update from Telegram
+// so that caller could get updates and send it not only there but to multiple sources
+func (t *Telegram) ProcessUpdate(ctx context.Context, textUpdate string) error {
```

Comment:
I think this t.run=true may cause some races, as Run probably will be run in a separate goroutine. At the same time public ProcessUpdate checks this flag to avoid mutually exclusive runs
maybe make it atomic?